kratos/includes/kratos_parameters.h

```diff
@@ -54,969 +51,1541 @@ namespace Kratos {
-
-          if (itr->value.IsObject()) {
-            if (!subobject.HasSameKeysAndTypeOfValuesAs(reference_subobject))
-              return false;
-          } else {
-            if (itr->value.GetType() != itr_ref->value.GetType()) {
-              /*std::stringstream msg;
-              msg << "The item with name :\"" << item_name << "\" does not have
-              the same type as the corresponding one in the default values" <<
-              std::endl;
-              msg << "The Parameters being validated are : " << std::endl;
-              msg << this->PrettyPrintJsonString() << std::endl;
-              msg << "Defaults against which the current parameters are
-              validated are :" << std::endl;
-              msg << reference.PrettyPrintJsonString() << std::endl;
-              KRATOS_THROW_ERROR(std::invalid_argument,"",msg.str());*/
-              return false;
+    ///@name Type Definitions
+    ///@{
+
+    /// Index definition
+    typedef std::size_t IndexType;
+
+    /// Size definition
+    typedef std::size_t SizeType;
+
+    /// Pointer definition of MmgProcess
+    KRATOS_CLASS_POINTER_DEFINITION(Parameters);
+
+    /// Definiton of the iterators
+    using iterator = iterator_adaptor;
+    using const_iterator = const_iterator_adaptor;
+
+    /// Iterators from nlohmann::json
+    typedef nlohmann::json::iterator json_iterator;
+    typedef nlohmann::json::const_iterator json_const_iterator;
+    typedef nlohmann::detail::iteration_proxy<json_iterator> json_iteration_proxy;
+    typedef nlohmann::detail::iteration_proxy<json_const_iterator> json_const_iteration_proxy;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * @brief Default constructor.
+     * @brief It assigns null pointers to the member variables
+     */
+    Parameters()
+    {
+        mpRoot = nullptr;
+        mpValue = nullptr;
+    }
+
+    /**
+     * @brief String constructor. It takes a string as input, which parses into a nlohmann::json class
+     * @param rJsonString The string to be parsed into a nlohmann::json class
+     */
+    Parameters(const std::string& rJsonString)
+    {
+        mpRoot = Kratos::shared_ptr<nlohmann::json>(new nlohmann::json( nlohmann::json::parse( rJsonString )));
+        mpValue = mpRoot.get();
+    }
+
+    /// Copy constructor.
+    Parameters(Parameters const& rOther)
+    {
+        //TODO: verify if mpValue is not null and eventually destruct correctly the data
+        mpRoot = rOther.mpRoot;
+        mpValue = rOther.mpValue;
+    }
+
+    /// Destructor.
+    virtual ~Parameters()
+    {
+//         delete[] mpValue;
+//         mpRoot = nullptr;
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    /// Assignment operator.
+    Parameters& operator=(Parameters const& rOther)
+    {
+        if(mpRoot.get() ==  mpValue || mpRoot == nullptr) {
+            mpRoot = Kratos::shared_ptr<nlohmann::json>(new nlohmann::json( nlohmann::json::parse( rOther.WriteJsonString() )));
```

Comment:
use make_shared here

Reply:
also to be honest i don't understand what is going on here...

Reply:
(I can change i, but the original implementation is yours)